src/gc.c

```diff
@@ -1108,6 +1116,8 @@ JL_DLLEXPORT jl_value_t *jl_gc_pool_alloc(jl_ptls_t ptls, int pool_offset,
         jl_gc_safepoint_(ptls);
     }
     gc_num.poolalloc++;
+    if(gc_statprofile_sample_rate && rand() < gc_statprofile_sample_rate)
```

Comment:
Following @timholy's comment on adjustible overhead: This implementation calls the RNG on every alloc. Hence, even if the sample rate is close to zero, the overhead does not converge to zero.

An alternative would be something like
`if(gc_num.poolalloc++ == gc_num.next_pool_sample) {gc_num.next_pool_sample += gc_statprofile_pool_inverse_rate;  jl_profile_record_trace(NULL);}`.

With `gc_num.next_pool_sample = 0`, this would trigger on next wrap-around, i.e. never, and with  `gc_statprofile_pool_inverse_rate` large this would trigger very rarely. We would pay only a single predicted branch on allocs we don't want to sample.

Similar treatment could be applied to separate gc_num.bigalloc, gc_num.allocd, etc counters. We probably should randomize the increment in order to avoid biases in loops that have period close to commensurable with the inverse rate. While poisson distribution of the gaps (as your code provides) is statistically nicer, something like `1 + (inverse_rate * rand_uint16()) >> 15` is probably good enough.

Reply:
That's a great point. I'll run some timings to see how RNG overhead compares to the allocation itself. If it's significant, I'll investigate the right scheme to use here. If not, there's probably value in keeping Poisson.